File: internal/httpapi/tickets.go
package httpapi

import (
	"io"
	"net/http"
	"strconv"

	"github.com/ArloB/tickets/internal/domain"
	"github.com/ArloB/tickets/internal/service"
)

type createTicketRequest struct {
	Type        string  `json:"type"`
	Title       string  `json:"title"`
	Description string  `json:"description"`
	Priority    string  `json:"priority"`
	Severity    *string `json:"severity"`
}

func (s *Server) createTicket(w http.ResponseWriter, r *http.Request) {
	projectKey := r.PathValue("key")

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req createTicketRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	fp, ferr := service.Fingerprint(r.Method, r.URL.Path, body)
	if ferr != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: ferr.Error()})
		return
	}

	var severity *domain.Severity
	if req.Severity != nil {
		sev := domain.Severity(*req.Severity)
		severity = &sev
	}

	ticket, err := s.svc.CreateTicket(r.Context(), service.CreateTicketRequest{
		ProjectKey:  projectKey,
		Type:        domain.TicketType(req.Type),
		Title:       req.Title,
		Description: req.Description,
		Priority:    domain.Priority(req.Priority),
		Severity:    severity,
	}, requestActor(r), correlationID(r), idempotencyKey(r), fp)
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusCreated, toTicketDetail(ticket))
}

// listTickets is GET /projects/{key}/tickets (product spec §5.6):
// priority-queue order by default, issue-register order via
// ?view=issue_register (§5.5). ?fields= narrows each list item to the
// named ticketCompact fields. Phase 4 added ?status=/?type=/?severity=
// /?priority=/?feature_ref=/?assignee=/?creator=/?updated_since=
// (docs/contracts/list-filters.md): each is optional and AND-composed
// with the others and with whichever ?view= selected the base
// ordering, and — like ?cursor= — must be resupplied on every page of
// a filtered listing (service.TicketListFilters' doc comment explains
// why the cursor doesn't encode them). The backlog view (product spec
// §6.1) is the default priority-queue ordering with filters layered
// on; there is no separate third ?view= value for it.
func (s *Server) listTickets(w http.ResponseWriter, r *http.Request) {
	projectKey := r.PathValue("key")
	q := r.URL.Query()

	limit := 0
	if v := q.Get("limit"); v != "" {
		n, err := strconv.Atoi(v)
		if err != nil || n < 0 {
			writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Field: "limit", Message: "limit must be a non-negative integer"})
			return
		}
		limit = n
	}

	filters := service.TicketListFilters{
		Status:       domain.WorkflowStatus(q.Get("status")),
		Type:         domain.TicketType(q.Get("type")),
		Severity:     domain.Severity(q.Get("severity")),
		Priority:     domain.Priority(q.Get("priority")),
		FeatureRef:   q.Get("feature_ref"),
		Assignee:     q.Get("assignee"),
		Creator:      q.Get("creator"),
		UpdatedSince: q.Get("updated_since"),
	}
	result, err := s.svc.ListTicketsFiltered(r.Context(), projectKey, service.TicketListView(q.Get("view")), limit, q.Get("cursor"), filters)
	if err != nil {
		writeError(w, r, err)
		return
	}
	compacted := make([]ticketCompact, len(result.Tickets))
	for i, t := range result.Tickets {
		compacted[i] = toTicketCompact(t)
	}

	fields := fieldNames(r)
	if len(fields) == 0 {
		writeJSON(w, http.StatusOK, ticketsPage{Tickets: compacted, NextCursor: result.NextCursor})
		return
	}
	if svcErr := validateFieldNames(allowedTicketCompactFields, fields); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	projected := make([]map[string]any, len(compacted))
	for i, c := range compacted {
		p, svcErr := projectFields(c, fields)
		if svcErr != nil {
			writeError(w, r, svcErr)
			return
		}
		projected[i] = p
	}
	writeJSON(w, http.StatusOK, map[string]any{"tickets": projected, "next_cursor": result.NextCursor})
}

// getTicket supports ?fields= (narrowing to the named ticketDetail
// fields) and ?include=comments,relationships (expanding the response
// with those sub-resources, docs/contracts/representations.md).
func (s *Server) getTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	ticket, err := s.svc.GetTicket(r.Context(), ref)
	if err != nil {
		writeError(w, r, err)
		return
	}
	detail := toTicketDetail(ticket)

	include := includeNames(r)
	if include["comments"] {
		comments, err := s.svc.ListComments(r.Context(), ref)
		if err != nil {
			writeError(w, r, err)
			return
		}
		out := make([]commentDetail, len(comments))
		for i, c := range comments {
			out[i] = toCommentDetail(c)
		}
		detail.Comments = &out
	}
	if include["relationships"] {
		out, err := relationshipViews(r.Context(), s.svc, ref)
		if err != nil {
			writeError(w, r, err)
			return
		}
		detail.Relationships = &out
	}

	fields := fieldNames(r)
	if len(fields) == 0 {
		writeJSON(w, http.StatusOK, detail)
		return
	}
	if svcErr := validateFieldNames(allowedTicketDetailFields, fields); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	projected, svcErr := projectFields(detail, fields)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	writeJSON(w, http.StatusOK, projected)
}

type updateTicketStatusRequest struct {
	Status string `json:"status"`
}

func (s *Server) updateTicketStatus(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req updateTicketStatusRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	ticket, err := s.svc.UpdateTicketStatus(r.Context(), service.UpdateTicketStatusRequest{
		Ref: ref, NewStatus: domain.WorkflowStatus(req.Status), ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

type updateTicketFieldsRequest struct {
	Type        string  `json:"type"`
	Title       string  `json:"title"`
	Description string  `json:"description"`
	Priority    string  `json:"priority"`
	Severity    *string `json:"severity"`
}

// updateTicketFields is PUT /tickets/{ref}: a full-representation
// update of every mutable ticket field except status (product spec
// §5.6, the plan's Step 13). Deliberately a distinct route from PATCH
// /tickets/{ref} rather than replacing it: that PATCH is the
// established, narrowly-scoped status transition endpoint
// (updateTicketStatus above), and merging the two would mean every
// drag-and-drop status change now has to also resend type/title/
// description/priority just to avoid clobbering them — PUT's "send
// the full representation" contract is the right fit for this
// operation, PATCH's "send the one thing that changed" contract is the
// right fit for that one.
func (s *Server) updateTicketFields(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req updateTicketFieldsRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	var severity *domain.Severity
	if req.Severity != nil {
		sev := domain.Severity(*req.Severity)
		severity = &sev
	}

	ticket, err := s.svc.UpdateTicketFields(r.Context(), service.UpdateTicketFieldsRequest{
		Ref: ref, Type: domain.TicketType(req.Type), Title: req.Title, Description: req.Description,
		Priority: domain.Priority(req.Priority), Severity: severity, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

type assignTicketRequest struct {
	Assignee *string `json:"assignee"`
}

func (s *Server) assignTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req assignTicketRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	var assignee *domain.ActorRef
	if req.Assignee != nil {
		parsed, perr := domain.ParseActorRef(*req.Assignee)
		if perr != nil {
			writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Field: "assignee", Message: perr.Error()})
			return
		}
		assignee = &parsed
	}

	ticket, err := s.svc.AssignTicket(r.Context(), service.AssignTicketRequest{
		Ref: ref, Assignee: assignee, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

type moveTicketRequest struct {
	Feature string `json:"feature"`
}

func (s *Server) moveTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req moveTicketRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	featureRef, svcErr := parseFeatureRef(req.Feature)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	ticket, err := s.svc.MoveTicketFeature(r.Context(), service.MoveTicketFeatureRequest{
		Ref: ref, NewFeatureRef: featureRef, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

type reorderTicketRequest struct {
	AfterRef *string `json:"after_ref"`
}

func (s *Server) reorderTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, r, &service.Error{Code: domain.ErrValidationFailed, Message: "failed to read request body"})
		return
	}
	var req reorderTicketRequest
	if svcErr := decodeJSON(body, &req); svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	var afterRef *domain.Reference
	if req.AfterRef != nil {
		parsed, svcErr := parseTicketRef(*req.AfterRef)
		if svcErr != nil {
			writeError(w, r, svcErr)
			return
		}
		afterRef = &parsed
	}

	ticket, err := s.svc.ReorderTicket(r.Context(), service.ReorderTicketRequest{
		Ref: ref, AfterRef: afterRef, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

func (s *Server) deleteTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	newVersion, err := s.svc.DeleteTicket(r.Context(), service.DeleteTicketRequest{
		Ref: ref, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, deleteResponse{Version: newVersion})
}

func (s *Server) restoreTicket(w http.ResponseWriter, r *http.Request) {
	ref, svcErr := parseTicketRef(r.PathValue("ref"))
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}
	version, svcErr := parseIfMatch(r)
	if svcErr != nil {
		writeError(w, r, svcErr)
		return
	}

	ticket, err := s.svc.RestoreTicket(r.Context(), service.RestoreTicketRequest{
		Ref: ref, ExpectedVersion: version,
	}, requestActor(r), correlationID(r))
	if err != nil {
		writeError(w, r, err)
		return
	}
	writeJSON(w, http.StatusOK, toTicketDetail(ticket))
}

func parseTicketRef(s string) (domain.Reference, *service.Error) {
	ref, err := domain.Parse(s)
	if err != nil {
		return domain.Reference{}, &service.Error{Code: domain.ErrValidationFailed, Field: "ref", Message: err.Error()}
	}
	if ref.Kind != domain.KindTicket {
		return domain.Reference{}, &service.Error{Code: domain.ErrValidationFailed, Field: "ref", Message: "reference must be a ticket reference"}
	}
	return ref, nil
}

File: web/src/api/tickets.ts
import { apiFetch, ifMatchHeader } from './client'
import type {
  Priority,
  Severity,
  TicketDetail,
  TicketsPage,
  TicketType,
  WorkflowStatus,
} from './types'

/** `view` selects the base ordering (§5.5/§5.6) — Milestone 2's
 * backlog view reuses priority_queue's order with filters layered on
 * (plan.md's Phase 4 plan, §4's ticket-list note). */
export type TicketListView = 'priority_queue' | 'issue_register'

export interface TicketListFilters {
  status?: WorkflowStatus
  type?: TicketType
  severity?: Severity
  priority?: Priority
  featureRef?: string
  assignee?: string
  creator?: string
  updatedSince?: string
}

export async function listTickets(
  projectKey: string,
  view: TicketListView,
  filters: TicketListFilters = {},
  cursor?: string,
  limit?: number,
): Promise<TicketsPage> {
  const params = new URLSearchParams({ view })
  if (filters.status) params.set('status', filters.status)
  if (filters.type) params.set('type', filters.type)
  if (filters.severity) params.set('severity', filters.severity)
  if (filters.priority) params.set('priority', filters.priority)
  if (filters.featureRef) params.set('feature_ref', filters.featureRef)
  if (filters.assignee) params.set('assignee', filters.assignee)
  if (filters.creator) params.set('creator', filters.creator)
  if (filters.updatedSince) params.set('updated_since', filters.updatedSince)
  if (cursor) params.set('cursor', cursor)
  if (limit) params.set('limit', String(limit))
  return apiFetch<TicketsPage>(`/projects/${encodeURIComponent(projectKey)}/tickets?${params}`)
}

/** `include` adds `comments`/`relationships` arrays to the detail
 * response — both keys are absent entirely unless requested
 * (docs/contracts/representations.md). */
export async function getTicket(
  ref: string,
  include: Array<'comments' | 'relationships'> = [],
  includeDeleted = false,
): Promise<TicketDetail> {
  const params = new URLSearchParams()
  if (include.length) params.set('include', include.join(','))
  if (includeDeleted) params.set('include_deleted', 'true')
  const query = params.toString()
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}${query ? `?${query}` : ''}`)
}

/** feature is the destination feature ref (e.g. "ABC-F2") — required by
 * the server, which no longer defaults to General (ADR 0023). The web
 * UI always sends a concrete ref rather than the server's `general`
 * shorthand, since it already has the full feature list to pick from. */
export interface CreateTicketInput {
  type: TicketType
  title: string
  description: string
  priority: Priority
  severity?: Severity | null
  feature: string
}

export async function createTicket(
  projectKey: string,
  input: CreateTicketInput,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/projects/${encodeURIComponent(projectKey)}/tickets`, {
    method: 'POST',
    body: input,
  })
}

export interface UpdateTicketFieldsInput {
  type: TicketType
  title: string
  description: string
  priority: Priority
  severity?: Severity | null
}

/** PUT /tickets/{ref} — full-representation update of every mutable
 * field except status; omitting a field clobbers it
 * (internal/httpapi/tickets.go's updateTicketFields doc comment). The
 * caller must resend every field from its base snapshot, not just the
 * ones the user touched. */
export async function updateTicketFields(
  ref: string,
  input: UpdateTicketFieldsInput,
  expectedVersion: number,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}`, {
    method: 'PUT',
    body: input,
    headers: ifMatchHeader(expectedVersion),
  })
}

/** PATCH /tickets/{ref} — status only, a separate flow from PUT above
 * with its own If-Match/version-conflict handling. */
export async function updateTicketStatus(
  ref: string,
  status: WorkflowStatus,
  expectedVersion: number,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}`, {
    method: 'PATCH',
    body: { status },
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function assignTicket(
  ref: string,
  assignee: string | null,
  expectedVersion: number,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}/assign`, {
    method: 'POST',
    body: { assignee },
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function moveTicketFeature(
  ref: string,
  featureRef: string,
  expectedVersion: number,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}/move`, {
    method: 'POST',
    body: { feature: featureRef },
    headers: ifMatchHeader(expectedVersion),
  })
}

/** POST /tickets/{ref}/reorder — move within a priority group.
 * `afterRef` must name a ticket in the *same* priority band
 * (api/openapi.yaml's ReorderRequest doc comment) — null/undefined
 * moves to the head of the group. A group renumber (when the gap
 * between neighbors is exhausted) bumps no sibling versions and emits
 * no audit event (ADR 0011), so callers never need to invalidate
 * other cached rows' versions after a successful reorder — only the
 * moved ticket's own response matters. */
export async function reorderTicket(
  ref: string,
  afterRef: string | null,
  expectedVersion: number,
): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}/reorder`, {
    method: 'POST',
    body: { after_ref: afterRef },
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function deleteTicket(ref: string, expectedVersion: number): Promise<{ version: number }> {
  return apiFetch<{ version: number }>(`/tickets/${encodeURIComponent(ref)}`, {
    method: 'DELETE',
    headers: ifMatchHeader(expectedVersion),
  })
}

export async function restoreTicket(ref: string, expectedVersion: number): Promise<TicketDetail> {
  return apiFetch<TicketDetail>(`/tickets/${encodeURIComponent(ref)}/restore`, {
    method: 'POST',
    headers: ifMatchHeader(expectedVersion),
  })
}
